src/Native/System.Native/pal_networking.cpp

```diff
@@ -379,7 +380,119 @@ static void ConvertHostEntPlatformToPal(HostEntry& hostEntry, hostent& entry)
     }
 }
 
-#if HAVE_GETHOSTBYNAME_R
+#if !HAVE_THREAD_SAFE_GETHOSTBYNAME_AND_GETHOSTBYADDR && (!HAVE_GETHOSTBYNAME_R && !HAVE_GETHOSTBYADDR_R)
+#define BUFFER_OVERFLOW 255
+static int copy_hostent(struct hostent* from, struct hostent* to,
+                        char* buffer, size_t buflen, int* error)
+{
+    int i, len;
+    char* ptr = buffer;
+
+    *error = 0;
+    to->h_addrtype = from->h_addrtype;
+    to->h_length = from->h_length;
+    to->h_name = static_cast<char*>(ptr);
+
+    /* copy hostname to buffer */
+    len = static_cast<int>(strlen(from->h_name)+1);
+    SafeStringCopy(ptr, len, from->h_name);
+    ptr += len;
+
+    /* copy aliases to buffer */
+    to->h_aliases = reinterpret_cast<char**>(ptr);
+    for(i = 0; from->h_aliases[i]; i++);
+    ptr += (i+1) * static_cast<int>(sizeof(char*));
+
+    for(i = 0; from->h_aliases[i]; i++)
+    {
+        len = static_cast<int>(strlen(from->h_aliases[i])+1);
+        if (static_cast<size_t>((ptr-buffer)+len) < buflen)
+        {
+            to->h_aliases[i] = ptr;
+            SafeStringCopy(ptr, len, from->h_aliases[i]);
+            ptr += len;
+        }
+        else
+        {
+            *error = BUFFER_OVERFLOW;
+            return *error;
+        }
+    }
+    to->h_aliases[i] = NULL;
+
+    /* copy addr_list to buffer */
+    to->h_addr_list = reinterpret_cast<char**>(ptr);
+    for(i = 0; reinterpret_cast<int*>(from->h_addr_list[i]) != 0; i++);
+    ptr += (i+1) * static_cast<int>(sizeof(int*));
+
+    for(i = 0; reinterpret_cast<int*>(from->h_addr_list[i]) != 0; i++)
+    {
+        len = sizeof(int);
+        if (static_cast<size_t>((ptr-buffer)+len) < buflen)
+        {
+            to->h_addr_list[i] = ptr;
+            memcpy(ptr, from->h_addr_list[i], static_cast<size_t>(len));
+            ptr += len;
+        }
+        else
+        {
+            *error = BUFFER_OVERFLOW;
+            return *error;
+        }
+    }
+    to->h_addr_list[i] = 0;
+    return *error;
+}
+
+static pthread_mutex_t lock_hostbyname_mutex = PTHREAD_MUTEX_INITIALIZER;
+static pthread_mutex_t lock_hostbyaddr_mutex = PTHREAD_MUTEX_INITIALIZER;
```

Comment:
The reason we need a lock is that the man pages state that gethostbyname and gethostbyaddr may return pointers to static data, aren't thread-safe, etc.  Do we know that they don't share such regions of memory?  If they possibly do, then they should be using the same single lock rather than two different ones.